server/src/main/java/org/apache/druid/segment/loading/SegmentLocalCacheManager.java

```diff
@@ -209,382 +278,173 @@ public void removeInfoFile(DataSegment segment)
-    synchronized (lock) {
+    final SegmentCacheEntryIdentifier identifier = new SegmentCacheEntryIdentifier(dataSegment.getId());
+    for (StorageLocation location : locations) {
+      final StorageLocation.ReservationHold<SegmentCacheEntry> hold =
+          location.addWeakReservationHoldIfExists(identifier);
       try {
-        File segmentDir = findStoragePathIfCached(segment);
-        if (segmentDir != null) {
-          return segmentDir;
+        if (hold != null) {
+          if (hold.getEntry().isMounted()) {
+            return new AcquireSegmentAction(
+                descriptor,
+                () -> Futures.immediateFuture(hold.getEntry().referenceProvider.acquireReference()),
+                hold
+            );
+          } else {
+            // go ahead and mount it, someone else is probably trying this as well, but mount is done under a segment
+            // lock and is a no-op if already mounted, and if we win we need it to be mounted
+            return new AcquireSegmentAction(
+                descriptor,
+                makeOnDemandLoadSupplier(dataSegment, hold.getEntry(), location),
+                hold
+            );
+          }
         }
-
-        return loadSegmentWithRetry(segment);
       }
-      finally {
-        unlock(segment, lock);
+      catch (Throwable t) {
+        throw CloseableUtils.closeAndWrapInCatch(t, hold);
       }
     }
-  }
-
-  /**
-   * If we have already reserved a location before, probably via {@link #reserve(DataSegment)}, then only that location
-   * should be tried. Otherwise, we would fetch locations using {@link StorageLocationSelectorStrategy} and try all
-   * of them one by one till there is success.
-   * Location may fail because of IO failure, most likely in two cases:<p>
-   * 1. druid don't have the write access to this location, most likely the administrator doesn't config it correctly<p>
-   * 2. disk failure, druid can't read/write to this disk anymore
-   * <p>
-   * Locations are fetched using {@link StorageLocationSelectorStrategy}.
-   */
-  private File loadSegmentWithRetry(DataSegment segment) throws SegmentLoadingException
-  {
-    String segmentDir = getSegmentDir(segment);
-
-    // Try the already reserved location. If location has been reserved outside, then we do not release the location
-    // here and simply delete any downloaded files. That is, we revert anything we do in this function and nothing else.
-    for (StorageLocation loc : locations) {
-      if (loc.isReserved(segmentDir)) {
-        File storageDir = loc.segmentDirectoryAsFile(segmentDir);
-        boolean success = loadInLocationWithStartMarkerQuietly(loc, segment, storageDir, false);
-        if (!success) {
-          throw new SegmentLoadingException(
-              "Failed to load segment[%s] in reserved location[%s]", segment.getId(), loc.getPath().getAbsolutePath()
+    final Iterator<StorageLocation> iterator = strategy.getLocations();
```

Comment:
Is it possible two callers start loading the same segment in two different locations? Presumably `strategy.getLocations()` might give different locations to different callers. And I don't see locking here.